### ISSUE TITLE
Cache issues with dtypes

### ISSUE THREAD

**Issue by TomAugspurger** [blaze/dask#551]
Not sure if this should be classified as user error.

```python
In [1]: import dask.dataframe as dd

In [2]: import pandas as pd

In [3]: import numpy as np

In [4]: df1 = pd.DataFrame({'A': [1, 2]})  # int dtype

In [5]: df2 = pd.DataFrame({'A': [np.nan, 3]})  # cast to float

In [6]: df1.to_csv('csv1.csv')

In [7]: df2.to_csv('csv2.csv')

In [8]: df = dd.read_csv('csv*.csv')

In [9]: df.sum().compute()
```

raises with a long traceback, including `cannot safely convert passed user dtype of <i8 for float64 dtyped data in column 1`.

This is just no integer-NA biting me, I can work around it by just reading in my data as floats I think.

**Comment by mrocklin** [comment]
Yeah, this came up recently with someone else.  We currently determine dtypes by looking at the first 1000 rows of the first CSV file.  This can and often does lead to incorrect results.

We used to look at many of the CSV files but this made `dd.read_csv` take a fairly long time on the pile of csvs case.  

I'm coming around to the idea that the right solution here is to catch the error at computation time and interpret it very clearly to the user  

*Hey, by looking at the first few lines of your first csv file we guessed that column: COLUMN had dtype DTYPE.  Now once we're actually reading all of the data we've found some values that don't match this dtype, probably you should add the following keyword argument to your `read_csv` call and compute again*

    dtype={'column-name':'dtype'}

Or something like that, possibly more concise.

**Comment by mrocklin** [comment]
What are some other solutions?

**Comment by TomAugspurger** [comment]
This is something I've ran into with pandas when streaming chunks for data from Redshift to HDF5. The best I've been able to do is what you've proposed, specifying the dtypes at IO time or `.astyping`.

The other option (in this case) is to try the computation, and then retry with the type upcast. I don't think dask should do this though, since it could lead to very surprising / hard to debug performance issues.
